pkg/client/patch.go

```diff
@@ -211,3 +212,34 @@ func (p applyPatch) Data(obj Object) ([]byte, error) {
 	// client-go does, more-or-less).
 	return json.Marshal(obj)
 }
+
+//applyFromPatch uses server-side-apply with a ApplyConfiguration object.
+type applyFromPatch struct {
+	patch interface{}
+}
+
+// Type implements Patch.
+func (p applyFromPatch) Type() types.PatchType {
+	return types.ApplyPatchType
+}
+
+// Sets the name and namespace for the returned object if an ApplyConfiguration was supplied
+func setNameNamespace(obj Object, ac interface{}) {
+	obj.SetName(reflect.ValueOf(ac).Elem().FieldByName("Name").Elem().String())
+	obj.SetNamespace(reflect.ValueOf(ac).Elem().FieldByName("Namespace").Elem().String())
```

Comment:
🤔 I'm surprised that `ObjectMetaApplyConfiguration` doesn't implement `metav1.Object` with getters and setters.

Reply:
Implementing the full `metav1.Object` doesn't make sense because other than Name and Namespace, most of the other fields are optional in an ApplyConfiguration while mandatory in an actual metav1.Object. Since the fields are mandatory in the metav1.Object, the interface prevents us from being able to return nil for fields. Eg: CreationTimestamp, Labels, etc.

We could technically implement a partial metav1.Object interface with just the getters and setters for Name and Namespace, but that'll require a change in k/k and push this code to be compatible with 1.22+ instead of 1.21+. 

cc @jpbetz

Reply:
> the interface prevents us from being able to return nil for fields. Eg: CreationTimestamp, Labels, etc.

I haven't thought this through, but I would expect a zero value from these when they are unset. The behavior should be like `metav1.ObjectMeta`, yes?

Looking closer, I see AC that have `*ObjectMetaApplyConfiguration` while their analogs have `metav1.ObjectMeta // +optional`... It's possible to call methods on a zero value of the latter, but calling methods on a `nil` pointer is not going to work out. Forget I asked about `metav1.Object`. 🙇🏻 

I believe this code needs to guard against nil ObjectMetaApplyConfiguration and nil Name and nil Namespace. I'm certain that `reflect.ValueOf(nilPointer).Elem().IsValid() == false`.

🤔 Do we need this at all? None of the other Patch implementations do this.

Reply:
So it turns out that `metav1.Object` is almost works out really well in apply configurations because all the fields are marked as 'omitempty' already. Honestly, the biggest problem with the `metav1.Object` is that it includes `ManagedFields` which is not allowed in apply configurations (you might notice if you look closely that `ObjectMetaApplyConfiguration` does not have `ManagedFields`).

Note that because we didn't implement `metav1.Object` on the kubernetes native types we've really can't require it in the Kubebuilder APIs (since being able to write controllers using Kubebuilder that applies the native types seems like a pretty reasonable requirement?). I'll admit this is a bit unfortunate.

Reply:
Overall, I don't think we should implement `metav1.Object` right now because (1) we don't implement it on the apply configurations for the kubernetes native types and (2) we exclude the `ManagedFields` field.